### ISSUE TITLE
Trained model into a pb file

### ISSUE THREAD

**Issue by gulingfengze** [GeorgeSeif/Semantic-Segmentation-Suite#105]
I want to turn the trained model into a pb file, but I don't know how to do it. Any good methods or suggestions?

**Comment by sercant** [comment]
Hey @gulingfengze, check out [this answer](https://stackoverflow.com/a/45868106/2542043) on stackoverflow.

**Comment by dtchuink** [comment]
Hi @gulingfengze 

Did you succeed on generating the pb file? I am still struggling to do it.

Thanks.

**Comment by gulingfengze** [comment]
@dtchuink  I'm sorry, I tried many times, but failed to generate pb file successfully.

**Comment by snehasusan2142** [comment]
I'm using front end - mobilenetv2 and model - mobileunet.  I was able to generate pb file from the checkpoint file (obtained from training) using the below code:


import tensorflow as tf

meta_path = 'MobileUNet_CamVid.ckpt.meta' # Your .meta file
output_node_names = ['Mean']    # Output nodes like softmax_cross_entropy_with_logits_sg:you may use tensorboard for this purpose

with tf.Session() as sess:

    # Restore the graph
    saver = tf.train.import_meta_graph(meta_path)

    # Load weights
    saver.restore(sess,tf.train.latest_checkpoint('/inputckpt/model'))#folder containing checkpoint

    # Freeze the graph
    frozen_graph_def = tf.graph_util.convert_variables_to_constants(
        sess,
        sess.graph_def,
        output_node_names)

    # Save the frozen graph
    with open('/MobileUNet_CamVid.pb', 'wb') as f:
      f.write(frozen_graph_def.SerializeToString())


But i wasnt able to do prediction using this pb file,code used for loading weights from checkpoint file is : 
saver=tf.train.Saver(max_to_keep=1000)
saver.restore(sess, args.checkpoint_path)
I think an equivalent of this is required to read pb file(i think i need to replace these lines after reading pb file using with gfile.FastGFile(args.pb_path,'rb') as f:)..

**Comment by snehasusan2142** [comment]
I was able to generate pb file and perform the inference with it.Output looks very good for me.
i used front end as Mobilenetv2 and SUPPORTED_MODEL mobileunet.
**Note:**while downloading the codes rename it to .py format to run in Python.

**Step1:**
During training you have to use an extra line after
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(logits=network, labels=net_output))
New line is:
new_network = tf.nn.softmax(network, name="softmax_output")#you can provide suitable output node name

**Step2:**
Generate checkpoint file. it should contain 4 files.
checkpoint,meta file,index file and a data file

**Step3:**
Generate Pb file using below code:ckpt2pb.py
[ckpt2pb.txt](https://github.com/GeorgeSeif/Semantic-Segmentation-Suite/files/3052621/ckpt2pb.txt)


**step4:**
You can optimize this graph using graph transformation tool:

bazel-bin/tensorflow/tools/graph_transforms/transform_graph \
--in_graph=generated_graph.pb \
--out_graph=optimized_graph.pb \
--inputs='Placeholder' \
--outputs='softmax_output' \
--transforms='
  add_default_attributes
  strip_unused_nodes(type=float, shape="1,299,299,3")
  remove_nodes(op=Identity, op=CheckNumerics)
  fold_constants(ignore_errors=true)
  fold_batch_norms
  fold_old_batch_norms
  quantize_weights
  quantize_nodes
  strip_unused_nodes
  sort_by_execution_order'


**step5:**
use this optimised pb file for inference using the below code:(slight modification made in the predict.py code):predictwithpb.py
[predictwithpb.txt](https://github.com/GeorgeSeif/Semantic-Segmentation-Suite/files/3052626/predictwithpb.txt)

with the help of tensorboard, we can visualize this graph.

**Comment by cena001plus** [comment]
@snehasusan2142  hello, i found one question, i kan convert ckpt mokel to pb model, but the pb model is small ,just 700k,  but the ckpt model is 10M ,  the ckpt model can give good result , but i load tflite  model , the result is wrong,    i can't find the reason. It's good if you can give me some sugestion.              @GeorgeSeif
